File: day_2/puzzle_1/solve.rb
# frozen_string_literal: true

require_relative '../../benchmark/benchmark'

Benchmark.run('day_2_puzzle_1') do
  play_hash = {
    'A' => { choice: :rock, beats: :scissors, score: 1 },
    'B' => { choice: :paper, beats: :rock, score: 2 },
    'C' => { choice: :scissors, beats: :paper, score: 3 },
    'X' => { choice: :rock, beats: :scissors, score: 1 },
    'Y' => { choice: :paper, beats: :rock, score: 2 },
    'Z' => { choice: :scissors, beats: :paper, score: 3 }
  }.freeze

  File.open('./day_2/input.txt', 'r') do |file|
    games = file.read.split("\n")

    score = games.inject(0) do |sum, game|
      oppenent_play = play_hash[game[0]]
      my_play       = play_hash[game[-1]]

      sum += my_play[:score]

      if my_play[:choice] == oppenent_play[:choice]
        sum += 3
      elsif my_play[:beats] == oppenent_play[:choice]
        sum += 6
      end

      sum
    end

    score
  end
end

File: day_2/puzzle_1/solve.py
import os
import sys
from functools import reduce
from typing import Dict, Literal, NamedTuple

sys.path.append(
    os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
)

from benchmark.benchmark import benchmark


def main():
    PlayKey = Literal["A", "B", "C", "X", "Y", "Z"]
    PlayName = Literal["rock", "paper", "scissors"]

    class Play(NamedTuple):
        choice: PlayName
        beats: PlayName
        score: int

    play_dict: Dict[PlayKey, Play] = {
        "A": Play("rock", "scissors", 1),
        "B": Play("paper", "rock", 2),
        "C": Play("scissors", "paper", 3),
        "X": Play("rock", "scissors", 1),
        "Y": Play("paper", "rock", 2),
        "Z": Play("scissors", "paper", 3),
    }

    def assess_game(sum: int, game: str):
        opponent_score = play_dict[game[0]]
        my_play = play_dict[game[-1]]

        sum += my_play.score

        if my_play.choice == opponent_score.choice:
            sum += 3
        elif my_play.beats == opponent_score.choice:
            sum += 6

        return sum

    with open("./day_2/input.txt", "r") as file:
        games = file.read().splitlines()
        score = reduce(assess_game, games, 0)
        return score


if __name__ == "__main__":
    benchmark(main, "day_2_puzzle_1")
